mixer/pkg/protobuf/yaml/decoder.go

```diff
@@ -0,0 +1,593 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yaml
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	multierror "github.com/hashicorp/go-multierror"
+)
+
+type (
+	// Decoder transforms protobuf-encoded bytes to attribute values.
+	Decoder struct {
+		resolver Resolver
+		fields   map[Number]*descriptor.FieldDescriptorProto
+	}
+)
+
+// NewDecoder creates a decoder specific to a dynamic proto descriptor.
+// Additionally, it takes as input an optional field mask to avoid decoding
+// unused field values. Field mask is keyed by the message proto field names.
+// A nil field mask implies all fields are decoded.
+// This decoder is specialized to a single-level proto schema (no nested field dereferences
+// in the resulting output).
+func NewDecoder(resolver Resolver, msgName string, fieldMask map[string]bool) *Decoder {
+	message := resolver.ResolveMessage(msgName)
+	fields := make(map[Number]*descriptor.FieldDescriptorProto)
+
+	for _, f := range message.Field {
+		if fieldMask == nil || fieldMask[f.GetJsonName()] || fieldMask[f.GetName()] {
+			fields[f.GetNumber()] = f
+		}
+	}
+
+	return &Decoder{
+		resolver: resolver,
+		fields:   fields,
+	}
+}
+
+// Decode function parses wire-encoded bytes to attribute values. The keys are field names
+// in the message specified by the field mask.
+func (d *Decoder) Decode(b []byte) (map[string]interface{}, error) {
+	visitor := &decodeVisitor{
+		decoder: d,
+		out:     make(map[string]interface{}),
+	}
+
+	for len(b) > 0 {
+		_, _, n := ConsumeField(visitor, b)
+		if n < 0 {
+			return visitor.out, ParseError(n)
+		}
+		b = b[n:]
+	}
+
+	return visitor.out, visitor.err
+}
+
+type decodeVisitor struct {
+	decoder *Decoder
+	out     map[string]interface{}
+	err     error
+}
+
+func (dv *decodeVisitor) setValue(f *descriptor.FieldDescriptorProto, val interface{}) {
+	if f.IsRepeated() {
+		var arr []interface{}
+		old, ok := dv.out[f.GetName()]
+		if !ok {
+			arr = make([]interface{}, 0)
```

Comment:
make([]interface, 0, 1), we know we are going to append to it right away.
Otherwise there is a needless potential alloc.

Reply:
Done